# Midterm study Notebook

The problems in Part One are taken from a previous quiz.  Write the value of each expression in the space provided. A helpful reminder.  When applied to strings, '<' checks alphabetical order.  So

```
>>> 'a' < 'b'
True
>>> 'b' < 'a'
False
>>> 'a' < 'a'
False
```

When applied to strings '+' does concatenation:

```
>>> 'a' + 'x'
'ax'
```


## Part One

In [3]:
lender_rates = {'Fannie Mae': .05, 'Bank of America': .065, 'Quicken Loans': .045}
prime_interest_rate = .04
amount_above_prime = [lender_rates[x] - prime_interest_rate for x in lender_rates.keys()]
loan_amounts = [100000, 150000, 200000, 500000]
interest_due = [x * prime_interest_rate for x in loan_amounts]
prime_mortgages = [x <= 200000 and x > 100000 for x in loan_amounts]
prime_mortgages2 = [x for x in loan_amounts if x <= 200000 and x > 100000]
shoe_string = 'adidas_ultra_boost'

In [4]:
(lender_rates, prime_interest_rate, amount_above_prime, 
 loan_amounts, interest_due, prime_mortgages)

({'Fannie Mae': 0.05, 'Bank of America': 0.065, 'Quicken Loans': 0.045},
 0.04,
 [0.010000000000000002, 0.025, 0.0049999999999999975],
 [100000, 150000, 200000, 500000],
 [4000.0, 6000.0, 8000.0, 20000.0],
 [False, True, True, False])

In [5]:
letters = 'abcdefghijklmnopqrstuvwxyz'
vowels = 'aeiou'
word = 'facetious'
X = [l for l in letters if l not in vowels]
Y = [x for x in letters if x > 'd' and x < 'j']
W = [x for x in letters if x == 'd' or x == 'j' or x < 'q']
Z = [x > 'd' and x < 'j' for x in letters]
R = [x for x in word if x in vowels]
consonants = [l for l in letters if l not in vowels]
#. This is a dictionary comprehension.  It creates a dictionary
consonant2vowel = {consonants[i]:v for (i, v) in enumerate(vowels)}
vowel2consonant = {v:k for (k,v) in consonant2vowel.items()}

# Exercises

# A

Write some code to turn `X` into a string. Remember `X` was created by a list comprehension, so it is initially defined to be a list.  The string you should get is:

```
'bcdfghjklmnpqrstvwxyz'
```

Solution:

In [5]:
''.join(X)

'bcdfghjklmnpqrstvwxyz'

# B

Write some code to produce a list of letters that are in `letters` but are not in `Y`.  Your code should use the current values of `letters` and `Y` and it should be a list comprehension using a filter.  Your result should look like this:

```
['a', 'b', 'c', 'd', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
```

Solution:

In [6]:
[l for l in letters if l not in Y]

['a',
 'b',
 'c',
 'd',
 'j',
 'k',
 'l',
 'm',
 'n',
 'o',
 'p',
 'q',
 'r',
 's',
 't',
 'u',
 'v',
 'w',
 'x',
 'y',
 'z']

# C

Write some code that returns a string consisting of the letters that are both in `W` and in `Y`.  

Solution:

In [7]:
[l for l in W if l in Y]

['e', 'f', 'g', 'h', 'i']

# D

Write a function called `find_vowels` that will take any string and return a **string** containing the vowels in that string.  For the vowels use  `vowels` (defined above)  After writing the function, you should
be able to apply it to `word` (as defined above).  The answer would be

```
>>> find_vowels(word)
'aeiou'
```

Then you should be able to
apply the function to `word` again and see whether the function returns a value `==` to `vowels`.
The result you should get is:

```
>>> find_vowels(word) == vowels
True
```

Notice that this last test also verifies that your function
returns a string, because only a string can be == to
a string.

You should be able to apply the function to `'radish'`.  The answer should be:

```
>>> find_vowels("radish")
'ai'
```

You should able to concatenate the answer the last question with
'x' as follows:

```
>>> find_vowels("radish") + 'x'
'aix'
```

Notice that this last test is a second way of verifying that your function
returns a string, because only a string can be concatenated with
a string.

Solution:

In [6]:
def find_vowels (inp = vowels):
    return ''.join([l for l in inp if l in vowels])

Comment:


On an in class midterm you won't be testing your code.  You'll just be writing your
code so that it handles all the test cases.  In the next three cells we demonstrate that the
code above does that.

In [7]:
find_vowels(word)

'aeiou'

In [8]:
find_vowels ('radish')

'ai'

In [9]:
find_vowels(word) == vowels

True

In [10]:
find_vowels("radish") + 'x'

'aix'

Comments:  you lose points if you wrote a function that printed something but has no
`return` statement.   You lose points if your code handles two test cases instead of all
three, and even more points if it handles one out of three, and still more if it handles
none.

You lose points if you return the wrong type of Python object.  The directions above clearly 
asked for a string.  If you put square brackets around your string, you are returning
a list of strings, not a string.  If you put curly braces around your string you are returning
a set of strings, not a string.  And so on.  Some of the above tests were devised so as to make you
think about this.  For example, consider the following bad function definition `find_vowels_bad`.

In [59]:
def find_vowels_bad (inp):
    return [''.join([l for l in inp if l in vowels])]

def find_vowels_bad2 (inp):
    print (''.join([l for l in inp if l in vowels]))

In [60]:
# Returns list of strings.  Bad.
find_vowels_bad ('radish')

['ai']

Fails this test since the output is a list, which cant == the string `vowels`.  Bad.

In [61]:
find_vowels_bad(word) == vowels

False

In [62]:
# Output cant be concatenated with string.  Bad.
find_vowels_bad ('radish') + 'x'

TypeError: can only concatenate list (not "str") to list

The function that prints its results fails two of the tests, as
predicted.  Its value is not == to `vowels` and its value
cannot be concatenated to `'x'`.

In [25]:
print(find_vowels_bad2('radish'))
print()
print(find_vowels_bad2(word)==vowels)
print()
print(find_vowels_bad2('radish')+'x')

ai
None

aeiou
False

ai


TypeError: unsupported operand type(s) for +: 'NoneType' and 'str'

# E

Modify `find_vowels` as defined in problem  **D** so that its argument is optional and it behaves as follows:

```
>>> find_vowels() == vowels
True
```

Note:  All the previously discussed examples using `find_vowels` should continue to behave as before.

Solution:

In [22]:
def find_vowels (inp = vowels):
    return ''.join([l for l in inp if l in vowels])

Comment:

In [19]:
find_vowels() == vowels

True

# F 

Now write a function `same_chars` that
determines if two sequences have exactly the same
set of characters.  It should behave as follows:

```
>>> same_chars(vowels, 'aeiou')
True
>>> same_chars(vowels, 'eioua')
True
>>> same_chars(vowels, 'uoiea')
True
>>> same_chars(vowels, 'eiou')
False
>>> same_chars('bcd','dcb')
True
>>> same_chars('bcd','dcb') == True
True
>>> if same_chars('xyz','zyx'):
...     print('Same!`)
... else:
        print('Different')
Same!
```

The last two tests verify that you are returning `True` or `False`, not justy printing them.
If you don't know what this means, study the functions nb and the functions assignment
solution.

Hint: Use sets. Remember how easy it is in Python to turn one type into another.  For example:

```
>>> list('abc')
['a', 'b', 'c']
```


Solution:

In [14]:
def same_chars (word1,word2):
    return set(word1)  == set(word2)

Comment:

In [15]:
same_chars('bcd','dcb')

True

In [17]:
same_chars(vowels, 'uoiea')

True

In [18]:
same_chars(vowels, 'uoie')

False

In [21]:
if same_chars('xyz','zyx'):
    print('Same!')
else:
    print('Different!')

Same!


# G

Consider `R` as defined above. Remember `R` filters the elements of `word` with the condition `x in vowels`.

Write some code that performs a test to show that `vowels` (a string) contains the same 
characters as `R` (a list).

Solution:

In [22]:
# Solution. The above definition of `same_chars` already does that.

same_chars(vowels,R)

True

# H

Consider `consonant2vowel`.  It is a mapping from consonants to vowels.  Write some code that returns the vowel associated with `"b"` in `consonant2vowel`.  Note:  This question has a **very** simple answer.  It is about
whether you understand what type of data structure `consonant2vowel` is, and how you use it.

Solution:

In [23]:
consonant2vowel['b']

'a'

# I

Write a simple one-line Python expression that operates on `vowels` and returns 

```
['ba', 'be','bi', 'bo','bu']
```

Solution:

In [24]:
['b' + x for x in vowels]

['ba', 'be', 'bi', 'bo', 'bu']

# J

Look at `interest_due`.  Using that as a model, write a simple expression that returns a list containing
the result of dividing every loan amount in `loan_amounts` by 100,000.


Solution:

In [27]:
[x/10**6 for x in loan_amounts]

# OK to write it out
#[x/100000 for x in loan_amounts]

[0.1, 0.15, 0.2, 0.5]

# K

Write a simple one line Python expression that operates on X and returns

```
['b', 'c']
```

Use slices.

Solution:

In [28]:
X[:2]

['b', 'c']

# L


Write a Python expression which operates on `letters` and returns the following value. 

```
'ghi'
```

Use slices.

Solution:

In [31]:
 letters[6:9]

'ghi'

# M

Write a Python expression which operates on `Z` and returns the following value

```
[False, False, False]
```

Solution:

In [33]:
Z[:3]

[False, False, False]

# N

Write a Python expression which operates on `Y` and returns the following value:

```
['e','i']
```

Solution:

In [35]:
[l for l in Y if l in vowels]

['e', 'i']

Comment:

Note `find_vowels` doesn't work because we defined it to return a string (as instructed).

In [26]:
find_vowels(Y)

'ei'

# O

Write a line of code which operates on `Y` and returns `False`.

Solution:

In [29]:
 len(Y)  == 6

False

# P

Write a line of code which accesses `vowels` and returns the following value:

```
[True, True, False, False, False]
```

Two of many solutions:

In [30]:
[x in 'ae' for x in vowels]

[True, True, False, False, False]

In [31]:
[x < 'i' for x in vowels]

[True, True, False, False, False]

# Q

Write a line of code which operates on `consonant2vowel` and returns the following value.

```
'e'
```

Solution:

`consonant2vowel` maps the first four consonants to the first four vowels:

In [44]:
consonant2vowel['c']

'e'



Write a line of code which operates on `vowel2consonant` and returns the following value.

```
'g'
```

Solution:

In [37]:
vowel2consonant['u']

'g'

##       Part Two

Write functions that behave according to the descriptions given.

# 1

Write a function that takes two numbers as arguments and multiplies them together and adds 1 to the result.  Call your function `product_successor`.  Then:

```
>>> product_successor(3,2)
7
>>> product_successor(5,9)
46
>>> product_successor(5,9) == 46
True
```

Last test verifies that the functioin is returning the value, not just printing it.

Solution:

In [41]:
def product_successor(x,y):
    return (x*y)+1

Comment:

In [42]:
product_successor(3,2)

7

In [43]:
product_successor(5,9)

46

In [107]:
product_successor(5,9) == 46

True

# 2

Write a function called `product_successor_ratios`
which takes four numbers as arguments.   Call them `a`,`b`,`c`' and `d`;
`product_successor_ratios` should first find the product successor of `a` and `b`.

That is, it should multiply `a` and `b` together
and add one to the result.  It should call the function `product_successor` (defined in the last problem))
to do this.  Note: **No credit for a solution that ignores this instruction**.   

`product_successor_ratio` should
also call `product_successor` on `c` and `d` (multiplying `c` and `d` together and adding 1 to the result).
It should return the ratio obtained by dividing the product successor of `a` and `b` 
by the product successor of `c` and `d`.

Note: It took a lot of English to say this but the body of the `product_successor_ratios`
function is essentially one line of code.  Here are some examples illustratinmg its behavior:

```
>>> product_successor_ratio(3,2,5,9)   # Test One
0.15217391304347827
>>> product_successor_ratio(5,9,3,2)   # Test Two
6.571428571428571
>>> product_successor_ratio(3,2,5,9) + product_successor_ratio(5,9,3,2)   # Test Three
6.72360248447205
>>> product_successor_ratio(5,9,3,0)    # Test Four
46.0
>>> product_successor_ratio(5,9,0,0)    # Test Five
46.0
>>> product_successor_ratio(5,9,0,0) == 46.0    # Test Five
True
```


Finally,  the last two arguments of `product_successor_ratio` should be
optional.  Here are the results when they are omitted.

```
>>> product_successor_ratio(5,9,1)
23.0
>>> product_successor_ratio(5,9)
23.0
```


Solution:

In [97]:
def product_successor_ratio (a,b,c=1,d=1):
    return product_successor(a,b)/product_successor(c,d)

Comments:

In [109]:
# Function from previous problem not called
def product_successor_ratio_no_credit (a,b,c=1,d=1):
    return ((a*b) + 1)/((c*d) + 1)

Tests

In [70]:
 product_successor_ratio(3,2,5,9) 

0.15217391304347827

In [111]:
product_successor_ratio(5,9,3,2) 

6.571428571428571

In [72]:
product_successor_ratio(3,2,5,9) + product_successor_ratio(5,9,3,2) 

6.72360248447205

In [73]:
product_successor_ratio(5,9,3,0)

46.0

In [74]:
product_successor_ratio(5,9,0,0)

46.0

In [113]:
#Also True 
#product_successor_ratio(5,9,0,0) == 46
product_successor_ratio(5,9,0,0) == 46.0

True

In [98]:
print(product_successor_ratio(5,9,1))

23.0


In [99]:
print(product_successor_ratio(5,9))

23.0


## Part 3

Reminders: `range(n)` returns a list of the integers from 0 up to but not including `n`.

```
>>> list(range(10))
[0,1,2,3,4,5,6,7,8,9]
```

`'str0'.join(str_seq)`  returns a string gotten by joining together the elements of `str_seq` with `str0`. It is the opposite of `split`.  It is an error if the elements of `str_seq` are not all strings:

```
>>> ''.join(['a','b','c'])
'abc'
>>> 'x'.join(['a','b','c'])
'axbxc'
>>> 'x'.join([1,2,3])
TypeError      
...
TypeError: sequence item 0: expected string, int found
>>> 'i'.join(['m','ss','ss','pp',''])
'mississippi'
>>> 'i'.join(['m','ss','ss','pp'])
'mississipp'
```




In [1]:
letters = ('a', 'b', 'c', 'd', 'e', 'f', 'g', 
           'h', 'i', 'j', 'k', 'l', 'm', 'n', 
           'o', 'p', 'q', 'r', 's', 't', 'u', 
           'v', 'w', 'x', 'y', 'z')
vowels = 'aeiou'
digits = list(range(4))
word_list = ['bbop','pine', 'bake', 'isle']
weird_list = [word_list[x][x] for x in digits]
word3 = ''.join([x[1] for x in word_list])
word4 = 'x'.join(['o','o','o',''])
char_list = [x for x in word4 if x in vowels]
word1 = 'rambunctious'
word2 = 'mellifluous'
WW = word1 + word2
Z = [x for x in WW if x in 'lmn']
favorite_words = {'Noam Chomsky': word2, 'Leonard Bloomfield': word1, 'John Rupert Firth': weird_list,
                 'Ferdinand de Saussure': word3}

# a

Write out an expression which applies to `word3` and returns

```
'bs'
```

Solution 1:

In [116]:
''.join([x for x in word3 if x not in vowels])

'bs'

Solution 2:

In [119]:
word3[::3]

'bs'

Comment:

In [120]:
 word3

'bias'

# b

Find a value for `LL` (a list)  for which the following piece of code produces the result shown.  There are some helpful examples above.

```
>>> 'a'.join(LL)
'abracadabra'
```

Solution:

In [100]:
LL=['','br','c','d','br','']
'a'.join(LL)

'abracadabra'

# c

What is the Python type of `digits` (as defined above)?

Solution

`list`.

# d

What is the Python type of `digits[0]`?

Solution:

`int`

Comment:

In [121]:
type(digits[0])

int

# e

Find a value for `word_list2` (a list) which makes the expression below have the given value
(see the example using `weird_list` above).

```
>>> [word_list2[x][x] for x in digits]
['v ','o','t','e']
```


Solution:

In [101]:
word_list2 = ['vixen','bot','bet','race']
[word_list2[x][x] for x in digits]

['v', 'o', 't', 'e']

Comment:

The hint says to look at `weird_list` so let's do that.

If we look at the value of `weird_list` (just evaluate it), it's

In [2]:
weird_list

['b', 'i', 'k', 'e']

Weird_list was created by

In [3]:
word_list = ['bbop','pine', 'bake', 'isle']
digits = list(range(4))
weird_list = [word_list[x][x] for x in digits]

Looking at digits we have

In [4]:
digits

[0, 1, 2, 3]

That means: to create `weird_list` with

```
weird_list = [word_list[x][x] for x in digits]
```

we collected the result of successively executing

```
word_list[0][0]  => 'bbop'[0] =>  'b'
word_list[1][1]  => 'pine[1] => 'i'
word_list[2][2]  => 'bake'[2] => 'k' 
word_list[3][3] =>  'isle'[3] =>  'e'
```

So to find word_list2 we need a similar list of 4 words  such  that 

```
word_list2[0][0]  => ??[0] =>  'v'
word_list2[1][1]  => ??[1] => 'o'
word_list2[1][0]  => ??[2] => 't' 
word_list[21][1] =>  ??[3] =>  'e'
```

There are lots of such word lists.  One
is given in the solution above.

# f

Write an expression which returns the **set** of vowels in  `char_list`.

Solution One:  use a **set comprehension**:

In [51]:
{x for x in char_list if x in vowels}

{'o'}

Solution Two:

In [52]:
set([x for x in char_list if x in vowels])

{'o'}

Solution Three

Since char_list consists **entirely** of vowels (by construction)

In [50]:
char_list

['o', 'o', 'o']

In [53]:
set(char_list)

{'o'}

also works.

# g 

Write an expression which operates on `favorite_words` and returns:

```
'ias'
```

Hint: Use a slice.  Remember that `favorite_word` is a dictionary. Values are retrieved from a dictionary by keys, and therefore it makes no sense to use numerical indices (or slices) directly on a dictionary. 

Solution:

In [129]:
favorite_words['Ferdinand de Saussure'][1:]

'ias'

# h

Write an expression which operates on `favorite_words` and returns:

```
'bun'
```


 Solution:

In [130]:
favorite_words['Leonard Bloomfield'][3:6]

'bun'

# i

Write an expression which retrieves

```
['a','b','c','x','y''z']
```

from `letters`.  Hint: use a list comprehension with a filtering condition.  This is not slice territory.

Solution

In [131]:
[x for x in letters if x < 'd' or x > 'w' ]

['a', 'b', 'c', 'x', 'y', 'z']

# j

What is the effect of or the value returned by the following code?  Explain
by referring to the type of `letters`.

In [48]:
letters[1]  = '0'

Solution

This is a `TypeError` because `letters` is a `tuple` and  a `tuple` is immutable and cannot have
a value assigned to it.

Comment:

In [132]:
letters[1]  = '0'

TypeError: 'tuple' object does not support item assignment

# k


What is the effect of or the value returned by the following code?  Explain
by referring to the type of `favorite_words`.

In [ ]:
favorite_words[0]  = 'Capitalism'

Solution:

This is a sucessful update which adds `0` as a new key to the `dict` `favorite_words`,
with the value `'Capitalism'`.

Comment:

Yes this is a bit of sneaky misdirection.  Dictionary updates will always work if the key is a legal
key for a dictionary (meaning the key is immutable), and `0` is a legal key (integers are immutable,
like strings).

Of course integers aren't usually used as keys for dictionaries beause if you really wanted to
access data by integer keys, your first thought would be to use a list.   But it might make sense
to use a dictionary in some cases, for example, if there are missing integers in the data, or
if the integer values will be supplied in some unpredicatble order.

In [133]:
favorite_words[0]  = 'Capitalism'

In [135]:
favorite_words

{'Noam Chomsky': 'mellifluous',
 'Leonard Bloomfield': 'rambunctious',
 'John Rupert Firth': ['b', 'i', 'k', 'e'],
 'Ferdinand de Saussure': 'bias',
 0: 'Capitalism'}

# l


What is the effect of or the value returned by the following code?  Explain
by referring to the type of `favorite_words`.

In [ ]:
favorite_words[2]

Solution:

This a `KeyError` since `2` is not a key in the `favorite_words` dictionary we have defined,
even if you count the successful update from the previous problem.

Comment:

In [136]:
favorite_words[2]

KeyError: 2

## Part 4

In [153]:
employees = ['happy', 'sleepy', 'dopey', 'bashful', 'grumpy', 'doc', 'sneezy']
employeeEmail = {'happy': ['sleepy','dopey'], 'dopey':['dopey'],'doc': employees,
                 'sleepy': ['bashful','doc', 'happy'],
                 'bashful': ['dopey']}
#employeeEmail[1] = ['norman']
secret_word = 'utilitarian'
flip = {'0':'1', '1':'0'}
bit_string = '001110'
flipped = [flip[x] for x in bit_string]

# 1

Write an expression which operates on `employees` and returns

```
['dopey', 'bashful', 'grumpy']
```

Solution

In [140]:
employees[2:5]

['dopey', 'bashful', 'grumpy']

# 2

Write an expression which operates on `employees` and returns

```
'y'
```

Solution:

In [143]:
employees[2][-1]

'y'

employeeEmail['happy'][1][-1]

# 3

Write an expression which operates on `employeeEmail` and returns

```
"rump"
```

Solution:

In [149]:
employeeEmail['doc'][4][1:5]

'rump'

The key here is the value assigned to the key `'doc'`, which
is the entire employees list. So 

```
employeeEmail = {'happy': ['sleepy','dopey'], 'dopey':['dopey'],'doc': employees,
                 'sleepy': ['bashful','doc', 'happy'],'grumpy':[],
                 'bashful': ['dopey']}
```
is in fact

In [146]:
employeeEmail

{'happy': ['sleepy', 'dopey'],
 'dopey': ['dopey'],
 'doc': ['happy', 'sleepy', 'dopey', 'bashful', 'grumpy', 'doc', 'sneezy'],
 'sleepy': ['bashful', 'doc', 'happy'],
 'grumpy': [],
 'bashful': ['dopey']}

# 4

Write an expression which operates on `secret_word` and returns

```
'arian'
```

Solution:

In [157]:
secret_word[-5:]

'arian'

# 5

Write an expression which operates on `secret_word` and returns

```
[False, False, True, False, True, False, True, False, True, True, False]
```

Solution  One:

In [162]:
[x in vowels[:-1] for x in secret_word]

[False, False, True, False, True, False, True, False, True, True, False]

Solution Two:

In [168]:
[x < 'l' for x in secret_word]

[False, False, True, False, True, False, True, False, True, True, False]

Comment:

The form of the solution is suggested by the fact the list of Booleans has
the same length as `secret_word`.

In [164]:
len([False, False, True, False, True, False, True, False, True, True, False])

11

In [165]:
len(secret_word)

11

# 6

`flipped` is defined above as

```
flipped = [flip[x] for x in bit_string]
```

The current value of `flipped` is

```
['1', '1', '0', '0', '0', '1']
```

Redefine `flip` so that the value of `flipped` becomes

```
[0, 0, 1, 1, 1, 0].
```


Solution One:

In [179]:
flipped = [int(flip[x]) for x in flipped]
flipped

[0, 0, 1, 1, 1, 0]

Solution Two:

In [176]:
flipped = [int(x) for x in bit_string]

[0, 0, 1, 1, 1, 0]

Solution Three:

In [174]:
[(int(x)+1)%2 for x in flipped]

[0, 0, 1, 1, 1, 0]

Part 5

In [205]:
import numpy as np
x = np.array([5,13,23,47,9])
y = np.ones((5,))
z = np.ones((5,2))
w = np.arange(15).reshape((3,5))

In [206]:
w

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

# a

Write out an expression which return the shape of `x`,

Solution

In [181]:
x.shape

(5,)

# b

Write out an expression which operates on `x` and returns

```
array([10, 26, 46, 94, 18])
```

Note: A solution using a `for`-loop or a list comprehension will receive no credit.

Solution One:

In [182]:
2*x

array([10, 26, 46, 94, 18])

# c

Write out an expression that returns an array containing the elementwise sum of the elements in `x` and `y`.

Solution:

In [188]:
x + y

array([ 6., 14., 24., 48., 10.])

# d

Consider:

Explain why you get a `ValueError` for `x + z`. 

Solution:

This is a ValueError because `x` and `z` have incompatible shapes for elementwise
arithmetic.  Nor can they  be broadcast into compatible shapes.

Comment:

In [185]:
x

array([ 5, 13, 23, 47,  9])

In [186]:
z

array([[1., 1.],
       [1., 1.],
       [1., 1.],
       [1., 1.],
       [1., 1.]])

In [190]:
x + z

ValueError: operands could not be broadcast together with shapes (5,) (5,2) 

This is one that **can** be fixed by reshaping.

In [189]:
x.reshape((5,1)) + z

array([[ 6.,  6.],
       [14., 14.],
       [24., 24.],
       [48., 48.],
       [10., 10.]])

In [192]:
x[:,np.newaxis] + z

array([[ 6.,  6.],
       [14., 14.],
       [24., 24.],
       [48., 48.],
       [10., 10.]])

# e

Note the following:

In [31]:
x < 20

array([ True,  True, False, False,  True], dtype=bool)

The result above is called a **Boolean array** because it only contains instances of the
type `bool` (`True` and `False`). Such Boolean arrays can be used to index arrays (basically
if they are of the same length).  Using a Boolean array this way is called using a **Boolean mask**.

# e (ctd.)

Using a Boolean mask, write an expression that returns an array containing the members of `x` that
are less than `20`.

Solution

In [193]:
x[x < 20]

array([ 5, 13,  9])

Comment:

In [194]:
x

array([ 5, 13, 23, 47,  9])

# f

Now consider

In [196]:
two_d_array = np.array([[1,2],[2,1],[5,5],[9,4]])
two_d_array

array([[1, 2],
       [2, 1],
       [5, 5],
       [9, 4]])

Write an expression that returns the second row of `two_d_array`.

Solution:

In [199]:
two_d_array[1,:]

array([2, 1])

Write an expression that returns `two_d_array` omitting the first row.

Solution:

In [200]:
 two_d_array[1:,:]

array([[2, 1],
       [5, 5],
       [9, 4]])

Write an expression that returns the following subarray of `two_d_array`.

```
array([[2, 1],
       [5, 5]])
```

Solution:

In [202]:
two_d_array[1:3,:]

array([[2, 1],
       [5, 5]])

In [204]:
x

array([ 5, 13, 23, 47,  9])

# g

Here is `w` again as a reminder. 

```
array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])
```

Use fancy indexing to retrieve the following array from `w`

```
array([ 2,  6, 13])
```

Note:   You answer should be one Python expression, and it should use fancy indexing. not Boolean
indexing.

Solution

In [208]:
w[[0,1,2],[2,1,3]]

array([ 2,  6, 13])

Comment

In [207]:
w

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

The values you want have the following indices in the 2D array:

```
[0, 2]
[1, 1]
[2, 3]
```

The row indices `[0`,`1`, `2]`, go on the left hand side of the comma.
The column indices `[2`,`1`, `3]`, go on the right hand side of the comma,
giving us


In [209]:
w[[0,1,2],[2,1,3]]

array([ 2,  6, 13])

# h

Below we load the `iris` data discussed in lecture and review some features:

In [211]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

In [212]:
features.shape

(150, 4)

Each row represents 4 measurements taken of an individual iris.

In [213]:
features[:10]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1]])

Various kinds of infromation anbout the dataset are included in `data`, a dictionary-like object.

In [214]:
data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

The measurements represented in the 4 columns are: 

In [215]:
data['feature_names']

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

So for example the petal length of the first iris (row 0 printed above) is 1.4.

The three species we are trying to classify are named in target names.

In [216]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

The array `target` (length 150) has the species for each iris in the data,
but represented as an integer.

In [217]:
set(target)

{0, 1, 2}

# f.1 

Your first task is to define a single Boolean array that is `True` for those rows of `features` that are not of target species `1`,
have sepal length greater than 4.7, and have petal width greater than 2.  Name your Boolean array `BC`.

Say what you think the length of  `BC` will be.

Solution:

In [232]:
BC = ((~ (target==1)) & (features[:,0] > 4.7) & (features[:,3] > 2))

The length of the `Boolean array` `BC` will be 150.

Comment:

In [237]:
len(BC)

150

The length of a Boolean array defined by some Boolean constraint on the columns of `features`,
or on an array of the same length as the columns (like `target`)  will always be the same length as the columns
of `features`.  In this case, since `features` has the shape `(150,4)`, that length is 150.

# f.2

Your next task is to write an expression that computes how many rows there are
that are not of target species `1`,
have sepal length greater than 4.7, and have petal width greater than 2.

Your expression should use `BC`. In fact, there are 23 such rows, so the value of your expression
should be 23.

Solution:

In [234]:
BC.sum()

Comment:

In [235]:
BC.sum()

23


Again a solution using `print(..)` has been ruled out.  The value of

```
print(BC.sum())
```

is `None`.

# f.3

Your final task is to write an expression that returns the rows of `features` that are not of target species 1, have sepal length greater than 4.7, and have petal width greater than 2.  It should use `BC`.

Solution:

In [225]:
features[BC]

array([[6.3, 3.3, 6. , 2.5],
       [7.1, 3. , 5.9, 2.1],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [7.2, 3.6, 6.1, 2.5],
       [6.8, 3. , 5.5, 2.1],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6.9, 3.2, 5.7, 2.3],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.1],
       [6.4, 2.8, 5.6, 2.2],
       [7.7, 3. , 6.1, 2.3],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.4, 2.1],
       [6.7, 3.1, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [6.8, 3.2, 5.9, 2.3],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5.2, 2.3],
       [6.2, 3.4, 5.4, 2.3]])

Comment:

As expected.

In [236]:
len(features[BC])

23

# f 4

Write an expression that computes how many of the rows computed in `f3` have target species `0`. (In fact it's `0`,
so your expression will have `0` as its value).

Solution:

In [ ]:
BC2 = ((features[:,0] > 4.7) & (target==0) & (features[:,3] > 2))
BC2.sum()

Comment:

There are three classes so the rows that don't belong to species 1 are either species 0 or species 2.
Apparently the sepal length and petal width measurements above rule out all irises of
species 0.

In [229]:
BC2.sum()

0

# f 5


Write another  expression that computes how many of the rows computed in `f3` have target value `2`. (In fact it's 23,
so your expression will have `23` as its value).

Solution:

In [230]:
BC3 = ((features[:,0] > 4.7) & (target==2) & (features[:,3] > 2))
BC3.sum()

Comment:

In [231]:
BC3.sum()

23

As expected, the sum of the species 0 and species 2 numbers equal the number we got when asked
for all irises not of species 1.